tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -260,48 +298,67 @@ instance GUnmarshal U1 where
 
 
 -- For unary products:
-instance (Selector s, Unmarshal k) => GUnmarshal (M1 S s (K1 c k)) where
-  gunmarshalNode _ = push getFields >>= gunmarshalProductNode . fromMaybe Map.empty
+instance UnmarshalAnn k => GUnmarshal (K1 c k) where
+  gunmarshalNode node = K1 <$> unmarshalAnn node
 
--- For sum datatypes:
-instance (GUnmarshalSum f, GUnmarshalSum g, SymbolMatching f, SymbolMatching g) => GUnmarshal (f :+: g) where
-  gunmarshalNode = gunmarshalSumNode @(f :+: g)
+-- For anonymous leaf nodes
+instance GUnmarshal Par1 where
+  gunmarshalNode node = Par1 <$> unmarshalAnn node
+
+instance (Unmarshal t, SymbolMatching t) => GUnmarshal (Rec1 t) where
+  gunmarshalNode node = Rec1 <$> unmarshalNode node
 
 -- For product datatypes:
 instance (GUnmarshalProduct f, GUnmarshalProduct g) => GUnmarshal (f :*: g) where
-  gunmarshalNode _ = push getFields >>= gunmarshalProductNode @(f :*: g) . fromMaybe Map.empty
-
-class GUnmarshalSum f where
-  gunmarshalSumNode :: (MonadFail m
-                   , Carrier sig m
-                   , Member (Reader ByteString) sig
-                   , Member (Reader (Ptr Cursor)) sig
-                   , MonadIO m)
-                   => Node -> m (f a)
+  gunmarshalNode node = push getFields >>= gunmarshalProductNode @(f :*: g) node . fromMaybe Map.empty
 
-instance (Unmarshal k, SymbolMatching k) => GUnmarshalSum (M1 C c (M1 S s (K1 i k))) where
-  gunmarshalSumNode node = M1 . M1 . K1 <$> unmarshalNodes [node]
-
-instance (GUnmarshalSum f, GUnmarshalSum g, SymbolMatching f, SymbolMatching g) => GUnmarshalSum (f :+: g) where
-  gunmarshalSumNode node = do
+-- For sum datatypes:
+instance (GUnmarshal f, GUnmarshal g, SymbolMatching f, SymbolMatching g) => GUnmarshal (f :+: g) where
+  gunmarshalNode node = do
     let lhsSymbolMatch = symbolMatch (Proxy @f) node
         rhsSymbolMatch = symbolMatch (Proxy @g) node
-    if lhsSymbolMatch
-      then L1 <$> gunmarshalSumNode @f node
-      else if rhsSymbolMatch
-        then R1 <$> gunmarshalSumNode @g node
-        else fail $ showFailure (Proxy @f) node `sep` showFailure (Proxy @g) node
+    if lhsSymbolMatch then
+      L1 <$> gunmarshalNode @f node
+    else if rhsSymbolMatch then
+      R1 <$> gunmarshalNode @g node
+    else
+      fail $ showFailure (Proxy @f) node `sep` showFailure (Proxy @g) node
 
 
 -- | Generically unmarshal products
 class GUnmarshalProduct f where
-  gunmarshalProductNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => Map.Map FieldName [Node] -> m (f a)
+  gunmarshalProductNode
```

Comment:
We still retain `GUnmarshalProduct` to deal with some of the quirks of products, e.g. field names being meaningful, and to avoid attempting to collect the fields for datatypes which don’t need them (where possible).